### PR TITLE
[BottomSheet] [3/3] Fix layout issues with bottom sheets containing scroll views

### PR DESCRIPTION
**Pull request by bhamiltoncx**
`MDCBottomSheetController` had a few subtle issues when the content view controller
contained a scroll view.

Contained scroll views always had their height set to the maximum
height of the `MDCSheetContainerView`, which might be larger than the
contained scroll view's content size.

With this PR, we ensure we never make the contained scroll view height
larger than the larger of (content size, desired sheet height).

Test Plan: Added new application-hosted tests to reproduce the
issues. I confirmed they failed before this PR and passed after. Ran
tests with:

  ./.kokoro --target //components/BottomSheet/...

### CODE REVIEW

File: material_components_ios.bzl [hunk 1]
```diff
@@ -95,7 +68,7 @@ def mdc_unit_test_suite(
     name = name,
     deps = deps,
     minimum_os_version = minimum_os_version,
-    runners = ios_runners(),
+    runners = DEFAULT_IOS_RUNNER_TARGETS,
```
Comment (jverkoey) [inline material_components_ios.bzl]: Love the testing cleanup - can you move it to a separate PR in case we need to roll the component changes back?

File: components/BottomSheet/BUILD [hunk 2]
```diff
@@ -45,6 +47,31 @@ mdc_objc_library(
...
+# test_host so we can exercise tests which depend on UIKit functionality like
+# UIViewController presentation logic.
+ios_unit_test_suite(
+    name = "app_tests",
+    deps = [":app_tests_lib"],
+    minimum_os_version = "8.0",
+    size = "medium",
+    test_host = "@build_bazel_rules_apple//apple/testing/default_host/ios",
+    runners = DEFAULT_IOS_RUNNER_TARGETS,
+)
```
Comment (jverkoey) [inline components/BottomSheet/BUILD:73]: Can we promote this to the .bzl file? Would be neat if other targets can reuse it.

Comment (jverkoey) [reply components/BottomSheet/BUILD:73]: Deleted prior comment down below because I think if you move this to the .bzl file we won't need to introduce the //components/testing/runners target.

**Review by jverkoey — CHANGES_REQUESTED** [review]
Please break the bazel testing changes out to a separate PR.

**Comment by bhamiltoncx** [comment]
Will split out. I'm not sure how to chain PRs so bear with me. :)

**Comment by jverkoey** [comment]
```
ERROR: /Volumes/BuildData/tmpfs/src/github/repo/components/BottomSheet/BUILD:50:1: C++ compilation of rule '//components/BottomSheet:app_tests_lib' failed (Exit 1)
components/BottomSheet/tests/app/MaterialBottomSheetAppTest.m:138:18: error: unrecognized platform name iOS
  if (@available(iOS 11.0, *)) {
                 ^
components/BottomSheet/tests/app/MaterialBottomSheetAppTest.m:139:16: error: property 'contentInsetAdjustmentBehavior' not found on object of type 'UIScrollView *'
    scrollView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentAlways;
               ^
components/BottomSheet/tests/app/MaterialBottomSheetAppTest.m:139:49: error: use of undeclared identifier 'UIScrollViewContentInsetAdjustmentAlways'
    scrollView.contentInsetAdjustmentBehavior = UIScrollViewContentInsetAdjustmentAlways;
                                                ^
3 errors generated.
```

**Comment by jverkoey** [comment]
Once you pull the testing target in to the .bzl file there might not be anything here worth pulling to a separate PR, so feel free to just clean up this PR in place.

**Comment by romoore** [comment]
Wouldn't the fix (and hosted unit test) still be worth having in this PR?

File: components/BottomSheet/src/private/MDCSheetContainerView.m [hunk 3]
```diff
@@ -239,6 +239,10 @@ - (void)updateSheetFrame {
     // If the content doesn't scroll then we have to set its frame to the size we are making
     // visible. This ensures content using autolayout lays out correctly.
     contentFrame.size.height = [self truncatedPreferredSheetHeight];
+  } else {
+    // If the content does scroll, don't make it larger than the content height or preferred height.
+    contentFrame.size.height =
+        MAX(self.sheet.scrollView.contentSize.height, [self truncatedPreferredSheetHeight]);
```
Comment (jverkoey) [inline components/BottomSheet/src/private/MDCSheetContainerView.m]: Should this be MIN, based on the comment?

**Review by jverkoey — COMMENTED** [review]
Yes — I think there probably isn't anything worth pulling out to a separate PR once the testing target has been added to the .bzl file.

**Comment by bhamiltoncx** [comment]
The trick with pulling the testing target into a separate PR is that the
ios_runners() function re-declares the runners as local targets every time
you call it.

So we can’t call it twice, which means we either introduce an implicit
ordering by calling it only for the first test suite and then referring to
the local runners, or we have to move the runners out into their own
targets.

That’s why I chose the latter.

**Comment by romoore** [comment]
I thought having the hosted app changes + runners changes would be sufficient to warrant a separate PR. That way if the code fixes break any clients, we can roll them back but keep the runner configurations?

**Comment by jverkoey** [comment]
> ios_runners() function re-declares the runners as local targets every time you call it.

Ah, that clarifies the need for the targets, thank you. Separate PR is probably best then to clean that up.

**Comment by jverkoey** [comment]
> I thought having the hosted app changes + runners changes would be sufficient to warrant a separate PR. That way if the code fixes break any clients, we can roll them back but keep the runner configurations?

Yes. It wasn't clear to me why we needed the runners target. Ben just pointed out that it's to avoid re-declaring the runners.

File: components/BottomSheet/src/private/MDCSheetContainerView.m (hunk 3 above)
Comment (jverkoey) [reply components/BottomSheet/src/private/MDCSheetContainerView.m]: I know it might be a pain but this change is potentially risky so it may be worth pulling out to a separate PR as well.

File: components/BottomSheet/BUILD (hunk 2 above)
Comment (bhamiltoncx) [reply components/BottomSheet/BUILD:73]: Split out into https://github.com/material-components/material-components-ios/pull/3422.

File: components/BottomSheet/src/private/MDCSheetContainerView.m [hunk 4]
```diff
@@ -234,11 +234,16 @@ - (void)updateSheetFrame {
   self.sheet.frame = sheetRect;
 
   CGRect contentFrame = self.sheet.bounds;
-  contentFrame.size.height -= kSheetBounceBuffer;
   if (!self.sheet.scrollView) {
     // If the content doesn't scroll then we have to set its frame to the size we are making
     // visible. This ensures content using autolayout lays out correctly.
     contentFrame.size.height = [self truncatedPreferredSheetHeight];
+  } else {
+    // If the content does scroll, make it the larger of the content
```
Comment (bhamiltoncx) [inline components/BottomSheet/src/private/MDCSheetContainerView.m]: Clarified comment.

**Comment by jverkoey** [comment]
Looks like this PR needs to be rebased.

**Comment by jverkoey** [comment]
The test runners are failing to run. Is it possible the bazel target changes somehow got into a bad state?

**Comment by jverkoey** [comment]
This PR breaks the bottom sheet example in MDCCatalog. Dragging up the bottom sheet does not result in it expanding as it should.

simulator screen shot - iphone 8 plus - 2018-05-22 at 17 19 16

**Comment by bhamiltoncx** [comment]
Thanks for finding. Must be a bug in the logic I added to MDCSheetContainerView.m.

How do I run that example, exactly?

**Comment by jverkoey** [comment]
Run MDC catalog, then run the bottom sheet primary demo.

**Comment by bhamiltoncx** [comment]
Thanks!

It looks like in that test, `BottomSheetDummyCollectionViewController` does not set `contentSize`, which leaves it at the default (`CGSizeZero`), which explains the crummy behavior after this PR.

[According to the docs](https://developer.apple.com/library/content/documentation/WindowsViews/Conceptual/UIScrollView_pg/CreatingBasicScrollViews/CreatingBasicScrollViews.html):

> You must set the contentSize property to the size of the scrollable content. This specifies the size of the scrollable area.

Since `BottomSheetDummyCollectionViewController` doesn't specify `contentSize`, that explains the problem.

Updating the contentSize like so in `-[BottomSheetDummyCollectionViewController viewDidLoad]` resolves this problem:

```
  self.collectionView.contentSize = CGSizeMake(
    self.view.bounds.size.width,
    (self.view.bounds.size.width / 3) * _numItems);
```

I've pushed to the PR to add this to the test.

In addition, I've updated this PR to disable the new logic entirely if `contentSize.height` is zero, and rebased the PR on master.

**Comment by jverkoey** [comment]
Just checked out the PR and it looks like the demo is still buggy. I believe it's because we're setting an explicit contentSize on a collection view, which will typically determine its own content size based on the data source.

If I remove the explicit contentSize code the demo does seem to work ok now though, so I'm not entirely sure how or why I was seeing the previous incorrect behavior now.

File: components/BottomSheet/examples/supplemental/BottomSheetDummyCollectionViewController.m [hunk 5]
```diff
@@ -43,6 +43,8 @@ - (void)viewDidLoad {
   [super viewDidLoad];
 
   self.collectionView.backgroundColor = [UIColor whiteColor];
+  self.collectionView.contentSize =
+      CGSizeMake(self.view.bounds.size.width, (self.view.bounds.size.width / 3) * _numItems);
```
Comment (jverkoey) [inline components/BottomSheet/examples/supplemental/BottomSheetDummyCollectionViewController.m:47]: Please revert this.

**Review by jverkoey — CHANGES_REQUESTED** [review]
Looks good. Please rebase and pull the UI changes out to a separate PR — unfortunately we can't enable UI tests at this point in time on our bazel presubmits.

**Comment by jverkoey** [comment]
This PR is actually blocking our a11y audit, so I'm going to close and re-open as a separate PR so that we can review and land asap.

Audit issue: https://github.com/material-components/material-components-ios/issues/3818
